src/Abstractions/TaskName.cs

```diff
@@ -8,127 +8,169 @@ namespace Microsoft.DurableTask;
-            this.Name = null!;
-            this.Version = null!;
-        }
-        else
-        {
-            this.Name = name;
-            this.Version = string.Empty; // expose setting Version only when we actually consume it.
-        }
+      // Force the default struct when null is passed in.
+      this.Name = null!;
+      this.Version = null!;
     }
-
-    /// <summary>
-    /// Gets the name of the task without the version.
-    /// </summary>
-    /// <value>
-    /// The name of the activity task without the version.
-    /// </value>
-    public string Name { get; }
-
-    /// <summary>
-    /// Gets the version of the task.
-    /// </summary>
-    /// <remarks>
-    /// Task versions is currently locked to <see cref="string.Empty" /> as it is not yet integrated into task
-    /// identification. This is being left here as we intend to support it soon.
-    /// </remarks>
-    public string Version { get; }
-
-    /// <summary>
-    /// Implicitly converts a <see cref="TaskName"/> into a <see cref="string"/> of the <see cref="Name"/> property value.
-    /// </summary>
-    /// <param name="value">The <see cref="TaskName"/> to be converted into a string.</param>
-    public static implicit operator string(TaskName value) => value.Name;
-
-    /// <summary>
-    /// Implicitly converts a <see cref="string"/> into a <see cref="TaskName"/> value.
-    /// </summary>
-    /// <param name="value">The string to convert into a <see cref="TaskName"/>.</param>
-    public static implicit operator TaskName(string value) => string.IsNullOrEmpty(value) ? default : new(value);
-
-    /// <summary>
-    /// Compares two <see cref="TaskName"/> objects for equality.
-    /// </summary>
-    /// <param name="a">The first <see cref="TaskName"/> to compare.</param>
-    /// <param name="b">The second <see cref="TaskName"/> to compare.</param>
-    /// <returns><c>true</c> if the two <see cref="TaskName"/> objects are equal; otherwise <c>false</c>.</returns>
-    public static bool operator ==(TaskName a, TaskName b)
+    else
     {
-        return a.Equals(b);
+      this.Name = name;
+      this.Version = string.Empty; // expose setting Version only when we actually consume it.
     }
+  }
 
-    /// <summary>
-    /// Compares two <see cref="TaskName"/> objects for inequality.
-    /// </summary>
-    /// <param name="a">The first <see cref="TaskName"/> to compare.</param>
-    /// <param name="b">The second <see cref="TaskName"/> to compare.</param>
-    /// <returns><c>true</c> if the two <see cref="TaskName"/> objects are not equal; otherwise <c>false</c>.</returns>
-    public static bool operator !=(TaskName a, TaskName b)
+  /// <summary>
+  /// Initializes a new instance of the <see cref="TaskName"/> struct.
+  /// </summary>
+  /// <param name="name">The name of the task. Providing <c>null</c> will yield the default struct.</param>
+  /// <param name="version">The version of the task.</param>
```

Comment:
Should we enforce a convention for the version? For example, it must either by an `Int32` or something that can be parsed by `System.Version.TryParse`? This will make it easier for us to implement automatic version selection on the client side, as well as make it easier for customers to do version comparisons in their code.

Reply:
I had thought about this before and I settled on not enforcing a convention, as I could see customers wanting a wide variety of versioning schemes. For example, Azure resources often version with a date format for their public API and they might want to use the same versioning scheme for durable.